python/google/protobuf/importer.py

```diff
@@ -0,0 +1,237 @@
+
+      def create_module(self, spec):
+        return None
+
+      def _generated_file_to_module_name(self, filepath):
+        components = filepath.split(os.path.sep)
+        return '.'.join(components[:-1] + [os.path.splitext(components[-1])[0]])
+
+      @staticmethod
+      def _register_message(sym_db, message_descriptor, module_name):
+        """Registers a message in the symbol database."""
+        nested_dict = {
+            '__module__': module_name,
+            'DESCRIPTOR': message_descriptor,
+        }
+        for nested_type_desc in message_descriptor.nested_types:
+          nested_type = ProtoLoader._register_message(sym_db, nested_type_desc,
+                                                      module_name)
+          nested_dict[nested_type_desc.name] = nested_type
+        for nested_enum_desc in message_descriptor.enum_types:
+          sym_db.RegisterEnumDescriptor(nested_enum_desc)
+        message_type = (
+            _reflection.GeneratedProtocolMessageType(message_descriptor.name,
+                                                     (_message.Message,),
+                                                     nested_dict))
+        sym_db.RegisterMessageDescriptor(message_descriptor)
+        sym_db.RegisterMessage(message_type)
+        return message_type
+
+      def exec_module(self, module):
+        """Instantiate a module identical to the generated version."""
+        file_descriptor = _pyext_message.FileDescriptor.FromFile(
+            self._protobuf_path.encode('ascii'),
+            [path.encode('ascii') for path in sys.path])
+        _sym_db.RegisterFileDescriptor(file_descriptor)
+        for dependency in file_descriptor.dependencies:
+          module_name = _proto_file_to_module_name(dependency.name)
+          if module_name not in sys.modules:
+            importlib.import_module(module_name)
+        setattr(module, 'DESCRIPTOR', file_descriptor)
+        for enum_name, enum_descriptor in file_descriptor.enum_types_by_name.iteritems(
+        ):
+          _sym_db.RegisterEnumDescriptor(enum_descriptor)
+          setattr(module, enum_name,
+                  enum_type_wrapper.EnumTypeWrapper(enum_descriptor))
+          for enum_value in enum_descriptor.values:
+            setattr(module, enum_value.name, enum_value.number)
+        for name, message_descriptor in file_descriptor.message_types_by_name.iteritems(
+        ):
+          message_type = ProtoLoader._register_message(_sym_db,
+                                                       message_descriptor,
+                                                       module.__name__)
+          _sym_db.RegisterMessageDescriptor(message_descriptor)
+          _sym_db.RegisterMessage(message_type)
+          setattr(module, name, message_type)
+
+    class ProtoFinder(importlib.abc.MetaPathFinder):
+      """MetaPathFinder that locates .proto files."""
+
+      def find_spec(self, fullname, path, target=None):
```

Comment:
What happens if there is both a *.proto file and a *_pb2.py file in the same directory?

Reply:
As you pointed out in the comment below, the "_pb2.py" file will take precedence. Let's close this discussion and continue it in that thread.